BndrunContainer incorrectly assumes artifactId == osgi.identity

During resolving with the Maven plugin a `-runrequires` is automatically calculated from the artifact id (https://github.com/bndtools/bnd/blob/3691481b9bde119f06deb3531e2c892a2d00d68c/biz.aQute.bnd.maven/src/aQute/bnd/maven/lib/resolve/BndrunContainer.java#L266). This must be the bundle symbolic name instead (https://docs.osgi.org/specification/osgi.core/7.0.0/framework.namespaces.html#i1776750)!

Otherwise you run into errors like
```
[INFO] --- bnd-resolver-maven-plugin:5.1.1:resolve (default) @ <artifact-id> ---
[INFO] Bnd inferred -runrequires: osgi.identity;filter:='(osgi.identity=<artifact-id>)'
[ERROR] Resolution failed. Capabilities satisfying the following requirements could not be found:
    [<<INITIAL>>]
      ⇒ osgi.identity: (osgi.identity=<artifact-id>)
```

Although this behaviour is documented at https://github.com/bndtools/bnd/tree/master/maven/bnd-resolver-maven-plugin#bndrun-details-inferred-from-maven it would be nice to improve that behaviour to evaluate the previously generated manifest instead.

--- a/biz.aQute.bnd.maven/src/aQute/bnd/maven/lib/resolve/BndrunContainer.java
+++ b/biz.aQute.bnd.maven/src/aQute/bnd/maven/lib/resolve/BndrunContainer.java
@@ -1,9 +1,12 @@
 package aQute.bnd.maven.lib.resolve;
 
+import static aQute.lib.exceptions.FunctionWithException.asFunction;
+
 import java.io.File;
 import java.util.Collections;
 import java.util.Enumeration;
 import java.util.List;
+import java.util.Map;
 import java.util.Objects;
 import java.util.Optional;
 import java.util.Properties;
@@ -26,6 +29,7 @@ import aQute.bnd.build.Workspace;
 import aQute.bnd.build.model.EE;
 import aQute.bnd.maven.lib.configuration.BeanProperties;
 import aQute.bnd.osgi.Constants;
+import aQute.bnd.osgi.Domain;
 import aQute.bnd.osgi.Processor;
 import aQute.bnd.repository.fileset.FileSetRepository;
 import aQute.bnd.service.RepositoryPlugin;
@@ -263,8 +267,22 @@ public class BndrunContainer {
 
 			Artifact artifact = project.getArtifact();
 
+			String bsn = artifact.getArtifactId();
+
+			try {
+				bsn = Optional.ofNullable(artifact.getFile())
+					.map(asFunction(Domain::domain))
+					.map(Domain::getBundleSymbolicName)
+					.map(Map.Entry::getKey)
+					.orElseGet(artifact::getArtifactId);
+			} catch (Exception e) {
+				logger.warn(
+					"Could not get the Bundle-SymbolicName from the project artifact {}, falling back to artifactId {}",
+					artifact, bsn, e);
+			}
+
 			run.setProperty(Constants.RUNREQUIRES,
-				String.format("osgi.identity;filter:='(osgi.identity=%s)'", artifact.getArtifactId()));
+				String.format("osgi.identity;filter:='(osgi.identity=%s)'", bsn));
 
 			logger.info("Bnd inferred {}: {}", Constants.RUNREQUIRES, run.getProperty(Constants.RUNREQUIRES));
 		}

--- a/maven/bnd-resolver-maven-plugin/README.md
+++ b/maven/bnd-resolver-maven-plugin/README.md
@@ -55,7 +55,7 @@ mvn bnd-resolver:resolve
 The `-runee` and `-runrequires` values can be inferred from the maven project as follows:
 
   * `-runee`, if omitted from the bndrun file, will be inferred from the `<target>` configuration of `maven-compiler-plugin`
-  * `-runrequires`, if omitted from the bndrun file, will be inferred from the project's `artifactId` and applied as `osgi.identity;filter:='(osgi.identity=<artifactId>)'`, if the project packaging is `jar` or `war` and the project has the `bnd-maven-plugin`
+  * `-runrequires`, if omitted from the bndrun file, will be inferred by attempting to get the `Bundle-SymbolicName` (bsn) from the project's main artifact and if not found will use `artifactId`. The value will be applied as `osgi.identity;filter:='(osgi.identity=<bsn|artifactId>)'`, if the project packaging is `jar` or `war` and the project has the `bnd-maven-plugin`
 
 ## Implicit Repository
 

--- a/maven/bnd-resolver-maven-plugin/src/test/resources/integration-test/test/pom.xml
+++ b/maven/bnd-resolver-maven-plugin/src/test/resources/integration-test/test/pom.xml
@@ -15,6 +15,7 @@
 		<module>resolve-from-scopes</module>
 		<module>resolve-from-inputbundles</module>
 		<module>enroute</module>
+		<module>resolve-infer-with-custom-bsn</module>
 	</modules>
 
 	<build>

--- a/maven/bnd-resolver-maven-plugin/src/test/resources/integration-test/test/resolve-infer-with-custom-bsn/pom.xml
+++ b/maven/bnd-resolver-maven-plugin/src/test/resources/integration-test/test/resolve-infer-with-custom-bsn/pom.xml
@@ -1,0 +1,71 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+	<modelVersion>4.0.0</modelVersion>
+
+	<parent>
+		<groupId>biz.aQute.bnd-test</groupId>
+		<artifactId>resolver-test</artifactId>
+		<version>0.0.1</version>
+	</parent>
+
+	<artifactId>resolve-infer-with-custom-bsn</artifactId>
+	<version>0.0.1</version>
+
+	<properties>
+		<maven.compiler.target>1.8</maven.compiler.target>
+	</properties>
+
+	<dependencies>
+		<dependency>
+			<groupId>org.apache.felix</groupId>
+			<artifactId>org.apache.felix.eventadmin</artifactId>
+			<version>1.4.8</version>
+		</dependency>
+		<dependency>
+			<groupId>org.apache.felix</groupId>
+			<artifactId>org.apache.felix.framework</artifactId>
+			<version>5.4.0</version>
+			<scope>runtime</scope>
+		</dependency>
+	</dependencies>
+
+	<build>
+		<plugins>
+			<plugin>
+				<groupId>biz.aQute.bnd</groupId>
+				<artifactId>bnd-maven-plugin</artifactId>
+				<configuration>
+					<bnd><![CDATA[
+						Bundle-SymbolicName: com.acme.foo
+					]]></bnd>
+				</configuration>
+				<executions>
+					<execution>
+						<goals>
+							<goal>bnd-process</goal>
+						</goals>
+					</execution>
+				</executions>
+			</plugin>
+			<plugin>
+				<groupId>biz.aQute.bnd</groupId>
+				<artifactId>bnd-resolver-maven-plugin</artifactId>
+			</plugin>
+			<plugin>
+				<artifactId>maven-compiler-plugin</artifactId>
+				<configuration>
+					<target>${maven.compiler.target}</target>
+				</configuration>
+			</plugin>
+			<plugin>
+				<artifactId>maven-jar-plugin</artifactId>
+				<configuration>
+					<archive>
+						<manifestFile>${project.build.outputDirectory}/META-INF/MANIFEST.MF</manifestFile>
+					</archive>
+				</configuration>
+			</plugin>
+		</plugins>
+	</build>
+</project>

--- a/maven/bnd-resolver-maven-plugin/src/test/resources/integration-test/test/resolve-infer-with-custom-bsn/src/main/java/com/acme/Foo.java
+++ b/maven/bnd-resolver-maven-plugin/src/test/resources/integration-test/test/resolve-infer-with-custom-bsn/src/main/java/com/acme/Foo.java
@@ -1,0 +1,5 @@
+package com.acme;
+
+public class Foo {
+	public static final Class<?> eventClazz = org.osgi.service.event.Event.class;
+}

--- a/maven/bnd-resolver-maven-plugin/src/test/resources/integration-test/test/resolve-infer-with-custom-bsn/test.bndrun
+++ b/maven/bnd-resolver-maven-plugin/src/test/resources/integration-test/test/resolve-infer-with-custom-bsn/test.bndrun
@@ -1,0 +1,1 @@
+-runfw: org.apache.felix.framework

--- a/maven/bnd-resolver-maven-plugin/src/test/resources/integration-test/test/verify.groovy
+++ b/maven/bnd-resolver-maven-plugin/src/test/resources/integration-test/test/verify.groovy
@@ -114,6 +114,33 @@ sb = new StringBuilder()
 bemRunBundles.get(0).formatTo(sb)
 assert sb.toString() == "org.apache.felix.eventadmin;version='[1.4.8,1.4.9)'"
 
+// The resolve-infer-with-custom-bsn case
+
+// Check the bndrun file exist!
+bndrunFile = new File(basedir, 'resolve-infer-with-custom-bsn/test.bndrun')
+assert bndrunFile.isFile()
+
+// Load the BndEditModel of the bndrun file so we can inspect the result
+processor = new Processor()
+processor.setProperties(bndrunFile)
+bem = new BndEditModel(Workspace.createStandaloneWorkspace(processor, bndrunFile.toURI()))
+doc = new Document(IO.collect(bndrunFile))
+bem.loadFrom(doc)
+
+// Get the -runbundles.
+bemRunBundles = bem.getRunBundles()
+assert bemRunBundles
+assert bemRunBundles.size() == 2
+
+sb = new StringBuilder()
+bemRunBundles.get(0).formatTo(sb)
+assert sb.toString() == "com.acme.foo;version='[0.0.1,0.0.2)'"
+sb = new StringBuilder()
+bemRunBundles.get(1).formatTo(sb)
+assert sb.toString() == "org.apache.felix.eventadmin;version='[1.4.8,1.4.9)'"
+
+
+/////////////////
 
 File build_log_file = new File("${basedir}/build.log")
 assert build_log_file.exists();

